clients/include/testing_spr.hpp

```diff
@@ -0,0 +1,191 @@
+
+        return;
+    }
+
+    size_t abs_incx = incx >= 0 ? incx : -incx;
+    size_t size_A   = N * (N + 1) / 2;
+    size_t size_x   = N * abs_incx;
+
+    // Naming: dK is in GPU (device) memory. hK is in CPU (host) memory
+    host_vector<T> hA_1(size_A);
+    host_vector<T> hA_2(size_A);
+    host_vector<T> hA_gold(size_A);
+    host_vector<T> hx(size_x);
+    host_vector<T> halpha(1);
+    CHECK_HIP_ERROR(hA_1.memcheck());
+    CHECK_HIP_ERROR(hA_2.memcheck());
+    CHECK_HIP_ERROR(hA_gold.memcheck());
+    CHECK_HIP_ERROR(hx.memcheck());
+    CHECK_HIP_ERROR(halpha.memcheck());
+
+    halpha[0] = h_alpha;
+
+    // allocate memory on device
+    device_vector<T> dA_1(size_A);
+    device_vector<T> dA_2(size_A);
+    device_vector<T> dx(size_x);
+    device_vector<T> d_alpha(1);
+    CHECK_HIP_ERROR(dA_1.memcheck());
+    CHECK_HIP_ERROR(dA_2.memcheck());
+    CHECK_HIP_ERROR(dx.memcheck());
+    CHECK_HIP_ERROR(d_alpha.memcheck());
+
+    double gpu_time_used, cpu_time_used;
+    double rocblas_gflops, cblas_gflops, rocblas_bandwidth;
+    double rocblas_error_1;
+    double rocblas_error_2;
+
+    // Initial Data on CPU
+    rocblas_init(hA_1, true);
+    rocblas_init(hx, false);
+
+    // copy matrix is easy in STL; hA_gold = hA_1: save a copy in hA_gold which will be output of
+    // CPU BLAS
+    hA_gold = hA_1;
+    hA_2    = hA_1;
+
+    // copy data from CPU to device
+    CHECK_HIP_ERROR(dA_1.transfer_from(hA_1));
+    CHECK_HIP_ERROR(dA_2.transfer_from(hA_1));
+    CHECK_HIP_ERROR(dx.transfer_from(hx));
+    CHECK_HIP_ERROR(d_alpha.transfer_from(halpha));
+
+    if(arg.unit_check || arg.norm_check)
+    {
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host));
+        CHECK_ROCBLAS_ERROR(rocblas_spr<T>(handle, uplo, N, &h_alpha, dx, incx, dA_1));
+
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_device));
+        CHECK_ROCBLAS_ERROR(rocblas_spr<T>(handle, uplo, N, d_alpha, dx, incx, dA_2));
+
+        // copy output from device to CPU
+        CHECK_HIP_ERROR(hA_1.transfer_from(dA_1));
+        CHECK_HIP_ERROR(hA_2.transfer_from(dA_2));
+
+        // CPU BLAS
+        cpu_time_used = get_time_us();
+        cblas_spr<T>(uplo, N, h_alpha, hx, incx, hA_gold);
+        cpu_time_used = get_time_us() - cpu_time_used;
+        cblas_gflops  = spr_gflop_count<T>(N) / cpu_time_used * 1e6;
+
+        if(arg.unit_check)
+        {
+            if(std::is_same<T, float>{} || std::is_same<T, double>{})
+            {
+                unit_check_general<T>(1, size_A, 1, hA_gold, hA_1);
+                unit_check_general<T>(1, size_A, 1, hA_gold, hA_2);
+            }
+            else
+            {
+                const double tol = N * sum_error_tolerance<T>;
+                near_check_general<T>(1, size_A, 1, hA_gold, hA_1, tol);
+                near_check_general<T>(1, size_A, 1, hA_gold, hA_2, tol);
+            }
```

Comment:
> 
> 
> I would put
> line 140: assert(std::is_same<T, rocblas_float_complex>{} || std::is_same<T, rocblas_double_complex>{});
> 
> It would help if anything ...

I would use `static_assert` instead of `assert`.

Reply:
If I'm not mistaken I can't use `static_assert` here unless this was an `if constexpr` statement, which we aren't using yet.
I've changed the order of the `if` statement to default to using `unit_check`s. I think this should solve any problems as if any new types are introduced we default to strict checking, rather than the more lenient `near_check`.